### ISSUE TITLE
CommunityService.updateCommunityLogo(file, uuid) does not work correctly

### ISSUE THREAD

**Issue by zipwiz** [OpenNTF/SocialSDK#1559]
Calls to CommunityService.updateCommunityLogo(file, communityUuid) do not set the given file as new community logo.

Additionally, with default settings a "logout" (LTPA token invalidation?) is forced.

For further details, see: 

http://stackoverflow.com/questions/24409152/ibm-sbt-communityservice-updatecommunitylogofile-communityuuid-triggers-logo

**Comment by SiriSch** [comment]
Is there a solution now? Have the same issue.

**Comment by zipwiz** [comment]
@SiriSch As I have written a year ago: With 1.0.2 it was buggy, but with 1.0.3 it was fixed. Thus I expect it to work still (with the current 1.1.12 or so), though I have not tested it again. Give it a try by updating your SBT version to the latest release build.

**Comment by SiriSch** [comment]
@zipwiz - the request itself seems to be ok but I will get redirected to the login page.
By this you can see that there are still some problems in the method, because the Communityservice works properly (for instance I can create a Communtiy). So there definitely must be a bug.

**Comment by zipwiz** [comment]
@SiriSch I will take a look in my working code the next few days, whether there is some kind of workaround I may have forgotten about. However my current code connects to IC5 inside a corporate intranet and authenticates by BasicAuth (special TAM role to allow that) with a "system user" because SSO behind the scenes (by using Forwardable LTPA token or similar) could not be established there, and I have not tried whether the new "runAs" request attribute also works for communitiy operations (don't think so by the way).

**Comment by zipwiz** [comment]
@SiriSch I have taken a look in my surrounding implementation and did not find any "workaround" or special handling. But currently the method is called only immediately after successful creation of a community, like "(0) build local community object; (1) create community; (2) load created community; (3) call updateCommunityLogo(logoFile, communityUuid)". And as I said I use HTTP BasicAuth and a "system user". The login page you receive may be caused by HTTP FormBased authentication which in turn could mean that the URL for "update logo" is not accessible by BasicAuth due to intranet security rules, which is a typical scenario in SSO environments. Try to check this with a browser-based REST client and make sure your browser session is NOT pre-authenticated by any means.

**Comment by SiriSch** [comment]
@zipwiz Firstly I have to say thanks that you took a view at your code!
I am calling the method as well after having created the community, so the community
definitely exists when I am calling the .updateCommuntiyLogo() method.

The intranet security rules could definitely be the problems causing this.

Currently I am doing the following:

```java
 public static boolean connect(String hostname, String user, String password) throws Exception {

    // setLoggingActive();

    Variables.hostname = hostname;
    Variables.username = user;
    Variables.password = password;
    Variables.baseURL = "https://" + Variables.hostname;
    Variables.origin = "https://" + Variables.hostname + "/";

    Variables.basicEndpoint.setUrl( Variables.baseURL );
    Variables.basicEndpoint.setUser( user );
    Variables.basicEndpoint.setPassword( password );
    Variables.basicEndpoint.setForceTrustSSLCertificate( true );
    Variables.basicEndpoint.setForceDisableExpectedContinue( true );
 

    try {
      Util.logit( ">> Login try with credentials:  " + user + "  " + password + " ..." );
      boolean loginOK = Variables.basicEndpoint.login( user, password );
      if (!loginOK) {
        Util.logit( ">> Login try with credentials:  " + user + "  " + password + "  has failed!" );
        return false;
      }
      Util.logit( ">> Login successfully done!" );
      Variables.referBasePoint();
    }
    catch (AuthenticationException e) {
      // TODO Auto-generated catch block
      e.printStackTrace();
    }
    return true;

  }


```

And the refering looks like this:

```java
public static void referBasePoint() {
    Variables.anActivityService = new ActivityService();
    Variables.anActivityService.setEndpoint( Variables.basicEndpoint );

    Variables.aBookMarkService = new BookmarkService();
    Variables.aBookMarkService.setEndpoint( Variables.basicEndpoint );

    Variables.aBlogService = new BlogService();
    Variables.aBlogService.setEndpoint( Variables.basicEndpoint );

    Variables.aCommunityService = new CommunityService();
    Variables.aCommunityService.setEndpoint( Variables.basicEndpoint );

    Variables.aFileService = new FileService();
    Variables.aFileService.setEndpoint( Variables.basicEndpoint );

    Variables.aForumService = new ForumService();
    Variables.aForumService.setEndpoint( Variables.basicEndpoint );

    Variables.aProfileService = new ProfileService();
    Variables.aProfileService.setEndpoint( Variables.basicEndpoint );

    Variables.aWikiService = new WikiService();
    Variables.aWikiService.setEndpoint( Variables.basicEndpoint );

  }

```

Did you set any other properties in relation to the basicEndPoint? And is this the way you use the basicEndPoint ?


Best regards


Sirius

**Comment by zipwiz** [comment]
@SiriSch I do not execute connect / login explicitly but trust to configurations. The SBT examples use managed-beans.xml and sbt.properties to configure different endpoints. This configuration is then automatically used when calling a service method like "createCommunity(...)" or "updateCommunityLogo(...)". Using this method lets your code automatically participate in the full authentication logic stack of the SBT instead of doing it "hand-crafted". Since that takes place anyway it may be that you run into conflicts between your hand-made connect/login code and the SBT stack. Try to solve it that way first -- at least check if you have a valid configuration set up properly.

**Comment by zipwiz** [comment]
@SiriSch An alternative is however indeed to build your own endpoint instance and feed it into the service constructors, just like you did. If you use Spring in your project, the endpoint Spring bean configuration could look like:

     <bean id="connectionsEndpoint" class="com.ibm.sbt.services.endpoints.ConnectionsBasicEndpoint">
         <property name="url" value="${connections.url}"/>
         <property name="user" value="${connections.user}"/>
         <property name="password" value="${connections.password}"/>
         <property name="apiVersion" value="4.5"/>
         <property name="authenticationPage" value="/sbt/loginForm.html?endpoint=connections"/>
         <property name="authenticationService" value="communities/service/atom/communities/my"/>
         <property name="forceTrustSSLCertificate" value="true"/>
         <property name="forceDisableExpectedContinue" value="true"/>
     </bean>

Basically this is the same as you have, but maybe the auth Page/Service is different (or unset on your side).

Having this in place with Spring, simply call getBean("connectionsEndpoint") to retrieve the endpoint to feed to the service instances (which also should be implemented as Spring beans if possible in that case).

**Comment by SiriSch** [comment]
@zipwiz 

Thanks for your effort!


I have replaced the login method that I have written myself.

The setting I haven't had checked was "Publish module contexts to separate XML files" ,this has to be checked in the TomCat Server Options.

I do not see any sense in using Spring - maybe you could tell me why you are using spring in sbt login context?

My managed-beans.mxl looks like that:

```xml
<?xml version="1.0"?>
<faces-config>
	<!-- Credential Store In-Memory Implementation -->
	<managed-bean>
		<managed-bean-name>CredStore</managed-bean-name>
		<managed-bean-class>com.ibm.sbt.security.credential.store.MemoryStore
		</managed-bean-class>
		<managed-bean-scope>application</managed-bean-scope>
	</managed-bean>

	<!-- Default Environment -->
	<managed-bean>
		<managed-bean-name>defaultEnvironment</managed-bean-name>
		<managed-bean-class>com.ibm.sbt.jslibrary.SBTEnvironment
		</managed-bean-class>
		<managed-bean-scope>application</managed-bean-scope>
		<managed-property>
			<property-name>endpoints</property-name>
			<value>connections</value>
		</managed-property>
	</managed-bean>

	<!-- CNX Endpoint -->
	<!-- Connections Basic -->
	<managed-bean>
		<managed-bean-name>connections</managed-bean-name>
		<managed-bean-class>com.ibm.sbt.services.endpoints.ConnectionsBasicEndpoint
		</managed-bean-class>
		<managed-bean-scope>session</managed-bean-scope>
		<managed-property>
			<property-name>url</property-name>
			<value>https://server1.blabla.com/</value>
		</managed-property>
		<managed-property>
			<property-name>credentialStore</property-name>
			<value>CredStore</value>
		</managed-property>
		<managed-property>
			<property-name>user</property-name>
			<value>THEUSERNAME</value>
		</managed-property>
		<managed-property>
			<property-name>password</property-name>
			<value>THEPASSWORD</value>
		</managed-property>
		<managed-property>
			<property-name>apiVersion</property-name>
			<value>4.5</value>
		</managed-property>
		<managed-property>
			<property-name>authenticationPage</property-name>
			<value>cce/loginForm.html?endpoint=connections</value>
		</managed-property>
		<managed-property>
			<property-name>authenticationService</property-name>
			<value>communities/service/atom/communities/my</value>
		</managed-property>
		<managed-property>
			<property-name>forceTrustSSLCertificate</property-name>
			<value>true</value>
		</managed-property>
		<managed-property>
			<property-name>forceDisableExpectedContinue</property-name>
			<value>true</value>
		</managed-property>
	</managed-bean>

	<!-- Password store based on DBCredentialStore -->
	<managed-bean>
		<managed-bean-name>CredStoreDB</managed-bean-name>
		<managed-bean-class>com.ibm.sbt.security.credential.store.DBCredentialStore
		</managed-bean-class>
		<!-- Managed Property jdbcDriverClass is required to provide the driver 
			class name, when jndi is not being used. The Property can be provided in 
			a way, mentioned below. -->
		<!-- <managed-property> <property-name>jdbcDriverClass</property-name> 
			<value>org.apache.derby.jdbc.EmbeddedDriver</value> </managed-property> -->
		<managed-property>
			<property-name>jndiName</property-name>
			<value>jdbc/ibmsbt-dbtokenstore</value>
		</managed-property>
		<managed-property>
			<property-name>tableName</property-name>
			<value>ADMINISTRATOR.SBTKREP</value>
		</managed-property>
		<managed-bean-scope>application</managed-bean-scope>
		<!-- Fixed OAuth Application parameters -->
	</managed-bean>
</faces-config>
```

The method .updateCommunityLogo() does not work either.

I have replaced the sbt.properties now, because I wanted to see whether this file is relevant but I cannot see any changes, so sbt.properties does not seem to be relevant.

The problem with updateCommunityLogo() so is still existing.

**Comment by zipwiz** [comment]
@SiriSch Thats bad. I only have some remarks and hints left:

Generally: It was just a try. The SBT has its own connect/login implementation and configuration, and usually it is a good idea to make use of what is offered in a context like this. But its no guarantee for things to work. And, of course, I am no SBT specialist -- I had just my own way of domesticating SBT for my project purposes, and it was not that easy at the beginning.

On Spring: Of course there is no need to use Spring here, but my project makes heavy use of Spring, thus it was straight forward to use Spring here also, allbeit for keeping configuration concerns together.

On sbt.properties: I have experienced similar things. Theoretically one should be able to specify ${expressions} in managed-beans.xml which are then resolved against sbt.properties entries, but I was not successful with that recipe. If everything is "hard-coded" in managed-beans.xml (and if the file is found and used) then sbt.properties may not be considered at all.

On versions: It may be of relevance to know whitch SBT version you currently use (1.0.x?, 1.1.x?, current = 1.1.11 / 1.1.12-SNAPSHOT?) and which Server version (4.0 / 4.5 / 5.0?) you connect to.

On the auth problem: If a "login page" is delivered instead of an operation result, then "form-based auth" is configured somewhere, and your application is confronted with that. But for things to work I think it is necessary to have BasicAuth in place (or OAuth or whatever, but a consistent protocol at least). If the server you talk to decides to switch from the intended auth method to form-based, then I suspect that there is a security configuration problem somewhere between your code and the server. To track that down it may be necessary to follow a complete conversation not only for failing requests like updateCommunityLogo() but also for successful requests like createCommunity(). Have an eye on the request / response headers and cookies: The LTPA token is a crucial thing for successful authentication and session management. (See the very start of this issue!)

On testing REST: If not already checked you should definitely try to resample your operations with a browser-based REST client, together with logging (and looking) what is wired. This will show you whats going on in the pure HTTP conversation. If ever possible you should try to bypass SSO when connecting from the browser to the server so as to require explicit authentication as part of the request or session: Usually your code will not be able to participate in SSO because it calls the server from behind the SSO scenes. (And if this is not true, then I want to know how you have managed that to happen.) Only if you are able to use forwardable LTPA tokens received before you will be able to participate in SSO.

**Comment by SiriSch** [comment]
@zipwiz 

Thanks!

I am using the newest version of sbt which currently is "1.1.11.20151208-1200".
The server version is 4.5..
